### ISSUE TITLE
136 -> 137 auto updater fails with unhandled exception

### ISSUE THREAD

**Issue by ntoff** [Rem0o/FanControl.Releases#1296]
**Describe the bug**
Auto update fails due to an unhandled exception, previous auto updates prior to this all functioned correctly.

From windows' log (no fan control logs are present)

```Application: Updater.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.Net.WebException
   at System.Net.HttpWebRequest.EndGetRequestStream(System.IAsyncResult, System.Net.TransportContext ByRef)
   at System.Net.Http.HttpClientHandler.GetRequestStreamCallback(System.IAsyncResult)

Exception Info: System.Net.Http.HttpRequestException
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
   at Updater.GoogleAnalytics+<Send>d__4.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
   at Updater.Program+<Main>d__7.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
   at Updater.Program.<Main>(System.String[])```

No "log.txt" file is present, this is an issue with the updater, fancontrol itself works perfectly fine.

**Comment by Exergist** [comment]
I also encountered an error (see below) when trying to update. I can't comment on prior versions because v136 was my first experience with FanControl. v136 is currently working fine for me. 

`Unhandled Exception: System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.WebException: The remote name could not be resolved: 'www.google-analytics.com'
   at System.Net.HttpWebRequest.EndGetRequestStream(IAsyncResult asyncResult, TransportContext& context)
   at System.Net.Http.HttpClientHandler.GetRequestStreamCallback(IAsyncResult ar)
   --- End of inner exception stack trace ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Updater.GoogleAnalytics.<Send>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Updater.Program.<Main>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Updater.Program.<Main>(String[] args)`
   
   There is no log file in the FanControl root directory.

**Comment by BlankFX1** [comment]
> I also encountered an error (see below) when trying to update. I can't comment on prior versions because v136 was my first experience with FanControl. v136 is currently working fine for me.

I get the exact same error too and it's obviously because the update isn't handling exceptions where people have a system wide block of google's spyware server on "www.google-analytics.com".
In my case it's blocked by a PiHole Instance running in our LAN so this domain can't get contacted.

I recommend to simply handle this exception and ignore it as google analytics isn't neccessary for the update process itself
OR - which might be the better choice - remove it completely as using it without the users explicit consense is illegal in at least all EU countries due to the GDPR.

**Comment by ntoff** [comment]
>In my case it's blocked by a PiHole Instance running in our LAN so this domain can't get contacted.

Same in my case.

Since this software is closed source, now I'm more than a little curious as to what other telemetry it collects without telling me.

**Comment by Exergist** [comment]
> In my case it's blocked by a PiHole Instance running in our LAN so this domain can't get contacted.
Same

**Comment by WoofWoofGrr** [comment]
Same issue, also running PiHole. I didn't notice the crash at first and my PC ended up shutting down due to overheating. I understand that bugs in software are unavoidable and I should have been paying more attention to make sure the update went fine. But finding out that unnecessary telemetry was what caused overheating to the point that it could have damaged my components has me reconsidering using this.

**Comment by Rem0o** [comment]
Alright here's the explanation:

Since at least a year or two, been using this bitly: https://bit.ly/3dO9JoF as a redirect for this url : https://raw.githubusercontent.com/Rem0o/FanControl.Releases/master/FanControl.zip, which downloads the latest version.

Why? It tells me the update count with a graph over time, that's it.  Gives me a sense of the userbase growth.  Hosting on Github is easy, but you get no sense of what's going on, unlike hosting on your own servers. 

Since I've recently created a new website for FanControl, I plugged some basic GA in there.

Thought the bitly thing was a botched solution at best, so might as well use one platform and replaced it with a GA4 event call with the downloaded version as its parameter, that's it. Same result as before, update count over time with a graph, now with the bonus of the version that was downloaded. I obviously didn't expect the DNS blocks and all that, that's my bad, should have expected this. Didn't handle that specific exception -> crash.

TLDR, this isn't me adding some new telemetry stuff, just replaced an existing update counter with a less shitty one ( or a shittier one, depending on who you ask I guess ).

For now I will put a catch and forget as a quick fix, so any firewall/pihole can/will block the call and the updater will move along.

I will review all of this, didn't know about the EU stuff, thanks for pointing that out, and I will think better solution probably with a basic EULA, I think we're at that point now. 

FYI, FanControl itself has no telemetry, and I don't plan on adding any.

**Comment by Deejay85** [comment]
While you removed the update from the releases page it is still asking to update whenever it is started. Not a real big issue, but one I thought you should know about. Also, unrelated, I use Batchrun (a GUI interface that requires less know how to run batch files) to launch a set of programs I use at startup, but for whatever reason, Fan Control won't run, even if I make a shortcut on the desktop, and link to that. Is there any way to make it so that it could be run from a batch file?

**Comment by Rem0o** [comment]
@Deejay85 

FanControl needs a delay when starting with Windows, otherwise the driver won't start properly.

**Comment by Exergist** [comment]
> Alright here's the explanation:
> 
> Since at least a year or two, been using this bitly: https://bit.ly/3dO9JoF as a redirect for this url : https://raw.githubusercontent.com/Rem0o/FanControl.Releases/master/FanControl.zip, which downloads the latest version.
> 
> Why? It tells me the update count with a graph over time, that's it. Gives me a sense of the userbase growth. Hosting on Github is easy, but you get no sense of what's going on, unlike hosting on your own servers.
> 
> Since I've recently created a new website for FanControl, I plugged some basic GA in there.
> 
> Thought the bitly thing was a botched solution at best, so might as well use one platform and replaced it with a GA4 event call with the downloaded version as its parameter, that's it. Same result as before, update count over time with a graph, now with the bonus of the version that was downloaded. I obviously didn't expect the DNS blocks and all that, that's my bad, should have expected this. Didn't handle that specific exception -> crash.
> 
> TLDR, this isn't me adding some new telemetry stuff, just replaced an existing update counter with a less shitty one ( or a shittier one, depending on who you ask I guess ).
> 
> For now I will put a catch and forget as a quick fix, so any firewall/pihole can/will block the call and the updater will move along.
> 
> I will review all of this, didn't know about the EU stuff, thanks for pointing that out, and I will think better solution probably with a basic EULA.
> 
> FYI, FanControl itself has no telemetry, and I don't plan on adding any.

So path forward is a new version that handles the exception in the short/medium term? Would reinstallation be required (since v136 cannot self-update)?

**Comment by Rem0o** [comment]
@Exergist  already reuploaded the Updater with the fix, shouldn't crash anymore, whatever if you block google servers or not.  You should just be able to update like before.

**Comment by Rem0o** [comment]
Closing as the bug itself is fixed. Still an update rework in on the radar though.
